Kernel/TermTransformer.hpp

```diff
@@ -47,18 +49,23 @@ class TermTransformer {
 };
 
 /**
- * TODO: WTF the name?
- *   --> rename to SumbtermEvaluator?
+ * Has similar philosophy to TermTransformer, but:
```

Comment:
I assume this stuff is just incidental cleanup unrelated to Twee per se. But thanks for doing it! Might be worth figuring out if `Kernel::BottomUpEvaluation" can be unified with this somehow.

Reply:
Exactly, according to the rule: if your PR touches something ugly, pay the price and make it nicer for rest of the codebase as well. ;)

A good question! @joe-hauns , you might know this. `TermTransformerTransformTransformed`, here being renamed to `BottomUpTermTransformer` is being used in, e.g., `InterpretedLiteralEvaluator`, which is now "your domain". What is it's relation to `BottomUpEvaluation` and why does "your" arithmetic code use both?

Reply:
So basically BUE (`BottomUpEvaluation`) does the same thing as the TTTT(`TermTransformerTransformTransformed`/`BottomUpTermTransformer`), but with a few performance optimizations. 
Say we have a term `f(t1...tn)`, and wanna bottom up apply the operation `φ` to it, then `TTTT` will compute `φ(t1) ...φ(tn)` first,  build the term `f(φ(t1)...φ(tn))`, and then compute `φ(f(φ(t1)...φ(tn)))`. 
BUE skips the step of creating `f(φ(t1)...φ(tn))`, and directly computes  `φ(f(φ(t1)...φ(tn)))`, which means we skip inserting some unnecessary intermediate terms in our term index.
Also BUE  not only works with terms but any sort of term-ish structure (i.e. DAGs). For example it's also used for z3 terms, and the arithmetically normalized terms `PolyNf`. 
Further BUE has some extras like never using dynamic dispatch, and cashing results in a hashmap, to not recompute results for the same subterms (which can be enabled using optional arguemnts).

The reason why both of them are used is that some of the old code I didn't write, but only updated uses TTTT. Everything I originally wrote uses BUE.

Reply:
So via a little code refactoring, we could get rid of TTTT altogether and use only BUE?

Reply:
That sounds like a win

Is there an easy correctness argument that can be made for BUE being able to replace TTTT directly?

Reply:
So it works as follows:
In TTTT we have 
```
Signature: TermList evalute(TermList)
Call: evaluate(f(φ(t1)...φ(tn)))
```
so `evaluate` computes `φ`.
In BUE we have 
```
Signature: TermList evalute(TermList old, TermList* evalutedArgs)
Call: evaluate(f(t1, ... tn), { φ(t1)...φ(tn) })
```
This means we do not directly compute phi, but we get symbol + arity from `old` and the arguments from `evaluatedArgs`.

Concerning the hashmap: 
Whenever we `evaluate(old, ...)` returns a result `r`, we put `(old, r)` in a hashmap, and before bottom-up evaluating some subterm, we check whether the result is already present in the map.

I can look into doing this refactor at some point soon, but I'm pretty busy implementing thing for POPL atm.

Reply:
* and i agree that it would be good to have a separate PR for that